crates/ruff_cli/src/commands/run.rs

```diff
@@ -80,18 +80,21 @@ pub(crate) fn run(
 
     // Load the caches.
     let caches = bool::from(cache).then(|| {
-        package_roots
-            .values()
-            .flatten()
-            .dedup()
-            .map(|package_root| {
-                let settings = resolver.resolve_all(package_root, pyproject_config);
+        let cache_roots = package_roots
+            .iter()
+            .map(|(package, package_root)| package_root.unwrap_or(package))
+            .unique()
+            .collect::<Vec<_>>();
+        cache_roots
+            .par_iter()
+            .map(|cache_root| {
+                let settings = resolver.resolve_all(cache_root, pyproject_config);
                 let cache = Cache::open(
                     &settings.cli.cache_dir,
-                    package_root.to_path_buf(),
+                    cache_root.to_path_buf(),
                     &settings.lib,
                 );
-                (&**package_root, cache)
+                (*cache_root, cache)
```

Comment:
I didn't spend any time optimizing this, feel free!

Reply:
(Not sure if worth calling `.unique()` above, if it's worth the `.collect()` allocation in order to allow the `.par_iter()`, if it's worth using the `.par_iter()` at all given far fewer entries, etc.)